File: 45-jump-game-ii/jump-game-ii.cpp
class Solution {
    int iter(int n, const vector<int>& nums) {
        vector<float> dp(n + 1, numeric_limits<float>::infinity());
        dp[1] = 0;

        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= nums[i - 1]; j++) {
                if (i + j <= n) {
                    dp[i + j] = min(dp[i] + 1, dp[i + j]);
                }
            }
        }
        return dp[n];
    }

public:
    int jump(vector<int>& nums) {
        return iter(nums.size(), nums);
    }
};

File: 45-jump-game-ii/jump-game-ii.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        def iter():
            n = len(nums)
            
            dp = [float('inf') for i in range(n+1)]
            dp[1] = 0
            for i in range(1, len(nums)+1):
                for j in range(1, nums[i-1]+1):
                    if i +j < n+1:
                        dp[i+j] = min( dp[i]+1, dp[i+j])
            return dp[n]
        return(iter())